pkg/health/server/server.go

```diff
@@ -0,0 +1,306 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"sync"
+	"time"
+
+	healthModels "github.com/cilium/cilium/api/v1/health/models"
+	healthApi "github.com/cilium/cilium/api/v1/health/server"
+	"github.com/cilium/cilium/api/v1/health/server/restapi"
+	ciliumModels "github.com/cilium/cilium/api/v1/models"
+	ciliumPkg "github.com/cilium/cilium/pkg/client"
+	"github.com/cilium/cilium/pkg/health/defaults"
+	"github.com/cilium/cilium/pkg/lock"
+
+	"github.com/go-openapi/loads"
+	flags "github.com/jessevdk/go-flags"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	log *logrus.Logger
+
+	// PortToPaths is a convenience map for access to the ports and their
+	// common string representations
+	PortToPaths = map[int]string{
+		defaults.HTTPPathPort: "Via L3",
+	}
+)
+
+// Config stores the configuration data for a cilium-health server.
+type Config struct {
+	Debug         bool
+	Passive       bool
+	Log           *logrus.Logger
+	CiliumURI     string
+	ProbeInterval time.Duration
+	ProbeDeadline time.Duration
+}
+
+// nodeMap maps IP addresses to NodeElements for convenient access to node
+// information.
+type nodeMap map[string]*ciliumModels.NodeElement
+
+// ConnectivityMap maps node names to their most recently seen status.
+type connectivityMap map[string]*healthModels.NodeStatus
+
+// Server is the cilium-health daemon that is in charge of performing health
+// and connectivity checks periodically, and serving the cilium-health API.
+type Server struct {
+	healthApi.Server  // Server to provide cilium-health API
+	*ciliumPkg.Client // Client to "GET /healthz" on cilium daemon
+	Config
+
+	waitgroup  sync.WaitGroup      // Used to synchronize all goroutines
+	tcpServers []*healthApi.Server // Servers for external pings
+	startTime  time.Time
+
+	// The lock protects against read and write access to the IP->Node map,
+	// the list of statuses as most recently seen, and the last time a
+	// probe was conducted.
+	lock.RWMutex
+	nodes        nodeMap
+	connectivity []*healthModels.NodeStatus
+	lastProbe    time.Time
+	localStatus  *healthModels.SelfStatus
+}
+
+// DumpUptime returns the time that this server has been running.
+func (s *Server) DumpUptime() string {
+	return time.Since(s.startTime).String()
+}
+
+// getNodes fetches the latest set of nodes in the cluster from the Cilium
+// daemon, and updates the Server's 'nodes' map.
+func (s *Server) getNodes() (nodeMap, error) {
+	scopedLog := logrus.NewEntry(log)
+	if s.CiliumURI != "" {
+		scopedLog = log.WithField("URI", s.CiliumURI)
+	}
+	scopedLog.Debug("Sending request for /healthz ...")
+
+	resp, err := s.Daemon.GetHealthz(nil)
+	if err != nil {
+		log.WithError(err).Warn("Failed to retrieve Cilium /healthz")
```

Comment:
In a k8s deployment, `cilium` gets deployed, which in turn starts up `cilium-health`, but `cilium` does not serve up its API before this query goes through, so we end up with this warning in the logs because `cilium-health` started up too early. We should perhaps look at deferring the `cilium-health` launch until after the API is served (See also #2116).

Reply:
Fixed.